# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/workers/WorkerPrivate.h
# Commit: f2debe329bca
# Full Hash: f2debe329bcaaad1248a234c165ccf240d602344
# Author: Eden Chuang <echuang@mozilla.com>
# Date: 2025-04-15 21:21:06
# Regressor Bug: 1899503
# File Overlap Count: 4
# Description:
#   Bug 1899503 - P3 PRemoteWorkerDebugger binding implementation. r=asuth
#   
#   This patch implements the PRemoteWorkerDebugger binding.
#   
#   We are trying to keep the same logic debugger registration/unregistration to minimize the difference between the local debugger(the debugger on the content process main thread)
# ==============================================================================

diff -r 604c4c52574a -r f2debe329bca dom/workers/WorkerPrivate.h
--- a/dom/workers/WorkerPrivate.h	Tue Apr 15 09:47:39 2025 +0000
+++ b/dom/workers/WorkerPrivate.h	Tue Apr 15 09:47:39 2025 +0000
@@ -65,7 +65,10 @@
 class ThrottledEventQueue;
 namespace dom {
 
+class PRemoteWorkerDebuggerChild;
+class PRemoteWorkerDebuggerParent;
 class RemoteWorkerChild;
+class RemoteWorkerDebuggerChild;
 class RemoteWorkerNonLifeCycleOpControllerChild;
 
 // If you change this, the corresponding list in nsIWorkerDebugger.idl needs
@@ -1061,6 +1064,24 @@
 
   void DisableDebugger();
 
+  void BindRemoteWorkerDebuggerChild();
+
+  void CreateRemoteDebuggerEndpoints();
+
+  void SetIsRemoteDebuggerRegistered(const bool& aRegistered);
+
+  void SetIsRemoteDebuggerReady(const bool& aReady);
+
+  void EnableRemoteDebugger();
+
+  void DisableRemoteDebugger();
+
+  void DisableRemoteDebuggerOnWorkerThread(const bool& aForShutdown = false);
+
+  void SetIsQueued(const bool& aQueued);
+
+  bool IsQueued() const;
+
   already_AddRefed<WorkerRunnable> MaybeWrapAsWorkerRunnable(
       already_AddRefed<nsIRunnable> aRunnable);
 
@@ -1135,7 +1156,7 @@
 
   void StartCancelingTimer();
 
-  const nsAString& Id();
+  const nsString& Id();
 
   const nsID& AgentClusterId() const { return mAgentClusterId; }
 
@@ -1497,6 +1518,15 @@
 
   mozilla::ipc::Endpoint<PRemoteWorkerNonLifeCycleOpControllerChild> mChildEp;
 
+  RefPtr<RemoteWorkerDebuggerChild> mRemoteDebugger;
+  mozilla::ipc::Endpoint<PRemoteWorkerDebuggerChild> mDebuggerChildEp;
+  mozilla::ipc::Endpoint<PRemoteWorkerDebuggerParent> mDebuggerParentEp;
+  bool mRemoteDebuggerRegistered MOZ_GUARDED_BY(mMutex);
+  bool mRemoteDebuggerReady MOZ_GUARDED_BY(mMutex);
+  bool mIsQueued;  // Should only touched on parent thread.
+  mozilla::CondVar mDebuggerBindingCondVar MOZ_GUARDED_BY(mMutex);
+  RefPtr<WorkerEventTarget> mWorkerDebuggerEventTarget;
+
   JS::UniqueChars mDefaultLocale;  // nulled during worker JSContext init
   TimeStamp mKillTime;
   WorkerStatus mParentStatus MOZ_GUARDED_BY(mMutex);